src/sb16.js

```diff
@@ -0,0 +1,1607 @@
+        this.dma.on_unmask(this.dma_channel_8bit, undefined);
+
+        var buffer = new Uint8Array(1);
+        buffer[0] = this.e2_value;
+        var syncbuffer = new SyncBuffer(buffer);
+
+        this.dma.do_read(syncbuffer, 0, 1, this.dma_channel_8bit, function()
+        {
+            dbg_log("e2 dma identification: written to dma", LOG_SB16);
+        });
+    });
+});
+
+// Get DSP copyright.
+register_dsp_command([0xE3], 0, function()
+{
+    this.read_buffer.clear();
+    for(var i = 0; i < DSP_COPYRIGHT.length; i++)
+    {
+        this.read_buffer.push(DSP_COPYRIGHT.charCodeAt(i));
+    }
+    // Null terminator.
+    this.read_buffer.push(0);
+});
+
+// Write test register.
+register_dsp_command([0xE4], 1, function()
+{
+    this.test_register = this.write_buffer.shift();
+});
+
+// Read test register.
+register_dsp_command([0xE8], 0, function()
+{
+    this.read_buffer.clear();
+    this.read_buffer.push(this.test_register);
+});
+
+// Trigger IRQ
+register_dsp_command([0xF2, 0xF3], 0, function()
+{
+    this.raise_irq();
+});
+
+// ASP - unknown function
+var SB_F9 = new Uint8Array(256);
+SB_F9[0x0E] = 0xFF;
+SB_F9[0x0F] = 0x07;
+SB_F9[0x37] = 0x38;
+register_dsp_command([0xF9], 1, function()
+{
+    var input = this.write_buffer.shift();
+    dbg_log("dsp 0xf9: unknown function. input: " + input, LOG_SB16);
+
+    this.read_buffer.clear();
+    this.read_buffer.push(SB_F9[input]);
+});
+
+
+
+//
+// Mixer Handlers
+//
+
+
+
+SB16.prototype.mixer_default_read = function()
+{
+    dbg_log("unhandled mixer register read. addr:" + h(this.mixer_current_address), LOG_SB16);
+    return this.mixer_unhandled_registers[this.mixer_current_address];
+}
+
+SB16.prototype.mixer_default_write = function(data)
+{
+    dbg_log("unhandled mixer register write. addr:" + h(this.mixer_current_address) + " data:" + h(data), LOG_SB16);
+    this.mixer_unhandled_registers[this.mixer_current_address] = data;
+}
+
+/**
+ * @param{number} address
+ * @param{function():number=} handler
+ */
+function register_mixer_read(address, handler)
+{
+    if(!handler)
+    {
+        handler = SB16.prototype.mixer_default_read;
+    }
+    mixer_read_handlers[address] = handler;
+}
+
+/**
+ * @param{number} address
+ * @param{function(number)=} handler
+ */
+function register_mixer_write(address, handler)
+{
+    if(!handler)
+    {
+        handler = SB16.prototype.mixer_default_write;
+    }
+    mixer_write_handlers[address] = handler;
+}
+
+// Reset.
+register_mixer_read(0x00, function()
+{
+    return 0;
+});
+register_mixer_write(0x00);
+
+// Output Stereo Select.
+register_mixer_write(0x0E, function(bits)
+{
+    this.dsp_stereo = bits & 0x2;
+    this.bus.send("speaker-stereo", this.dsp_stereo);
+    this.bus.send("speaker-filter", bits & 0x20);
+});
+
+// IRQ Select.
+register_mixer_read(0x80, function()
+{
+    switch(this.irq)
+    {
+        case 2: return 0x1;
+        case 5: return 0x2;
+        case 7: return 0x4;
+        case 10: return 0x8;
```

Comment:
Can you make constant for the irqs?

Reply:
Like a constant array that maps the irq values directly to the return values? In that case, would you like me to do the same for the subsequent functions up to line 1235? (DMA Select and IRQ status)

Reply:
No, I was thinking of using `case SB16_INT_FOO:` instead of the magical values.

What you're suggesting might also be a good idea, but I wouldn't do that if the array is sparse.

Reply:
Ah I see what you mean.

Reply:
Hmmm, the best that I can come up with is `SB16_IRQ2 = 2`, `SB16_IRQ5 = 5`, etc, because the SB16 device only uses one IRQ line for everything. In the perspective of the device, these different IRQ lines have no special meaning.